snuba/subscriptions/scheduler_processing_strategy.py

```diff
@@ -0,0 +1,150 @@
+from collections import deque
+from datetime import datetime
+from enum import Enum
+from typing import Deque, Mapping, Optional, cast
+
+from arroyo import Message
+from arroyo.processing.strategies import ProcessingStrategy
+
+from snuba.subscriptions.utils import Tick
+from snuba.utils.metrics import MetricsBackend
+
+
+class SchedulerMode(Enum):
+    IMMEDIATE = "immediate"
+    WAIT_FOR_SLOWEST_PARTITION = "wait-for-slowest"
+
+
+class TickBuffer(ProcessingStrategy[Tick]):
+    """
+    The TickBuffer buffers ticks until they are ready to be submitted to
+    the next processing step.
+
+    The behavior of the TickBuffer depends on which of the two scheduler
+    modes applies.
+
+    If the scheduler mode is IMMEDIATE then there is no buffering and a
+    message is always immediately submitted to the next processing step.
+
+    If the scheduler mode is WAIT_FOR_SLOWEST_PARTITION then messages are
+    buffered until all partitions are at least up to the timestamp of the
+    tick.
+
+    `max_ticks_buffered_per_partition` applies if the scheduler mode is
+    WAIT_FOR_SLOWEST_PARTITION. Once the maximum ticks is received for that
+    partition, we start to submit ticks for processing even if that timestamp
+    is not received for all partitions yet.
+
+    This prevents the buffered tick list growing infinitely and scheduling
+    to grind to a halt if one partition starts falling far behind for some reason.
+    """
+
+    def __init__(
+        self,
+        mode: SchedulerMode,
+        partitions: int,
+        max_ticks_buffered_per_partition: Optional[int],
+        next_step: ProcessingStrategy[Tick],
+        metrics: MetricsBackend,
+    ) -> None:
+        if mode == SchedulerMode.WAIT_FOR_SLOWEST_PARTITION:
+            assert max_ticks_buffered_per_partition is not None
+
+        self.__mode = mode
+        self.__partitions = partitions
```

Comment:
I think the subject of topic and partition management is actually quite separate to subscriptions, and there are other impacts that are not just about subscriptions, so I don't know that this is easily resolved here.

Today the way of dealing with Kafka topics in Snuba is a bit different depending on the environment.

In on-premise, single tenant and dev at least, the topic configuration comes from Snuba, and the topic gets created with the number of partitions specified via settings during bootstrap. The Snuba setting TOPIC_PARTITION_COUNTS acts as the source of truth for the partition configurations. This is the assumption on which the scheduler works today, the current subscriptions also works, and other parts of our codebase do.

The implication of your suggestion seems to be that Snuba shouldn't be responsible for managing topics at all, it should just read live configuration from Kafka on startup or even periodically and use that anywhere it needs to know the partition count. Basically someone else should be responsible for managing these configurations.

I think there are good arguments for pursuing something like what you suggest but we've never really discussed it before and it seems quite different to how our codebase works today. So I am pretty hesitant to go ahead and build something just for subscriptions that is different from how we do things elsewhere. I think there is really no pressing need or immediate benefit from adding this to subscriptions immediately, and it's kind of tangential to how the scheduler works.

Anyway it seems to me if we go down this route of using configuration from the admin API then we will also want to make some other changes to Snuba including deprecating settings.TOPIC_PARTITION_COUNTS. At that point it would probably make sense to just have `KafkaTopicSpec.partitions_number` (which in turn checks settings today) handle fetching the value and not necessarily even changing the scheduler implementation at all since that already calls that same function.

Ultimately I see this as a mostly separate problem to subscription scheduling, since it might affect other things as well. Until there is a clearer long term direction I would prefer to keep using `KafkaTopicSpec.partitions_number` not because I'm advocating for doing it this way longer term but mostly because I'd rather preserve the status quo rather than introduce a new way of doing things in one place, at least until we have some more agreement on the overall direction.

Either way, I don't think it makes any difference to this strategy as it will most likely always get passed a partition count in the constructor from someone else outside who will be doing the fetching.

Reply:
I agree this may not impact the strategy (wrote it above).
I forgot about bootstrap, though I have some vague memories about us discussing whether we should have a separate set of parameters for initialization of a topic (for bootstrap) from a way to access the current states of some parameters.

In practice, you are right, the situation now is ambiguous. We do not make a distinction between the two cases.
I think the root cause is exactly this. There should be an initial configuration somewhere (settings can be ok), but then Kafka settings are generally modified outside of snuba and snuba needs to adapt to those changes instead of driving them. At least this is what would happen today in all environment as we do not provide a way (through snuba) to manage the topics after creation.

If Anton needed to add partitions to the events topic in prod today, he would use the kafka console and snuba would know nothing about it, which would be disastrous. 
Let's ensure we fix this process in the next quarter. While adding partitions to events is reasonably unlikely, other topics we are adding alerts onto are more likely to change the partitions number.